release-content/0.16/release-notes/gpu_driven_rendering.md

```diff
@@ -0,0 +1,75 @@
+
+* *GPU transform and cull*, which allows Bevy to compute the position and visibility of every object from the camera's perspective on the GPU instead of on the GPU.
+
+* The *retained render world*, which allows the CPU to avoid processing and uploading data that hasn't changed since the last frame. Bevy 0.16 not only retains entities in the render world from frame to frame but also retains many internal rendering data structures, allowing it to avoid most CPU overhead associated with meshes that didn't change from the previous frame.
+
+Not all platforms offer full support for GPU-driven rendering; for example, Metal (macOS and iOS) currently only has partial support, and WebGL 2 (the browser) has no support for GPU-driven rendering at all. Some platforms, such as Android phones with certain Mali chips, theoretically support GPU-driven rendering, but in practice driver issues prevent it from working reliably. In these cases, Bevy 0.16 automatically works around the missing functionality, or falls back to CPU-driven rendering if that isn't feasible. This is possible because the Bevy renderer is a unified codebase that supports both CPU- and GPU-driven rendering techniques. That is, there is just one renderer, not a CPU-driven renderer and a GPU-driven renderer.
+
+In most cases, you don't need to do anything special in order for your application to support GPU-driven rendering. There are two main exceptions:
+
+1. Materials with custom WGSL shaders will continue to use CPU-driven rendering by default. In order for your materials to use GPU-driven rendering, you'll want to use the new `#[bindless]` feature on `AsBindGroup`. See the `AsBindGroup` documentation and the `shader_material_bindless` example for more details. If you're using `ExtendedMaterial`, check out the new `extended_material_bindless` example.
+
+2. Applications and plugins that hook into the renderer at a low level will need to be updated to support GPU-driven rendering. The newly-updated `custom_phase_item` and `specialized_mesh_pipeline` examples may prove useful as a guide to do this.
+
+Bevy's current GPU-driven rendering isn't the end of the story. There's a sizable amount of potential future work to be done:
+
+* Bevy 0.16 only supports GPU-driven rendering for the 3D pipeline, but the techniques are equally applicable to the 2D pipeline. Future versions of Bevy could support 2D GPU-driven rendering as well.
+
+* Morph targets currently force objects to CPU-driven rendering. This is something we plan to address in the future.
+
+* We're watching new API features, such as [Vulkan device generated commands] and [Direct3D 12 work graphs], with interest. These would allow future versions of Bevy to offload even more work to the GPU, such as sorting of transparent objects. While figuring out how to unify these disparate APIs in a single renderer will be a challenge, the future possibilities in this space are exciting.
```

Comment:
You could already do transparent object sorting on the GPU. Write out a buffer of depths and then run a radix sort or something, and then do indirect draws.

What DGC and work graphs really do is stuff like only bind and run the pipeline/shader for materials actually on screen (skipping materials for entities that were entirely culled).

Reply:
That only works if all your transparent objects use the same shader, which isn't realistic for most apps. The reason you need DGC for real world transparent sorting is that you don't know when to switch shaders until you've done the sorting.

Reply:
Ah yeah this is true.